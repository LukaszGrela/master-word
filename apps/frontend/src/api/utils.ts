import { TPartialGameState } from './types';

export const createGameState = (
  attempts: number,
  wordLength: number
): TPartialGameState => {
  const word = Array.from(Array(wordLength)).map(() => '');
  return Array.from(Array(attempts)).map(() => ({
    word: word.concat(),
  }));
};
