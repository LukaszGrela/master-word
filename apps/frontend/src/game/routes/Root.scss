@import '../../styles/variables';

.master-word {
  .footer {
    .gd-logo {
      height: 2rem;
      width: 2.5625rem;

      @media (prefers-color-scheme: dark) {
        fill: $gd-red-dark;
      }
    }
  }
}
