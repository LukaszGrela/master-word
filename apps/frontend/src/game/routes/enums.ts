export enum EPaths {
  ROOT = '/',
  GAME = '/game',
  RESULTS = '/results/:result',
}

export const getResultsPath = (result: 'win' | 'lose') => `/results/${result}`;
