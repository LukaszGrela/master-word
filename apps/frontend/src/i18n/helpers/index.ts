/* v8 ignore start */
export * from './loadTranslation';
/* v8 ignore stop */
