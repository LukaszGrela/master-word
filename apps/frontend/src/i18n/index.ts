/* v8 ignore start */
export * from './LanguageContext/hooks';
export { LanguageProvider } from './LanguageContext/index';
export type { ILanguageContext } from './LanguageContext/index';
/* v8 ignore stop */
