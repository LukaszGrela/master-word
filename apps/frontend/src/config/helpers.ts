import { AppStorage, EStorageKeys } from '@repo/utils';
import { IConfig } from './types';
import { ATTEMPTS, LANGUAGE, WORD_LENGTH } from '../game';

/**
 * Assure local storage is using enabled values from current config object
 * @param config Current Configuration
 */
export function updateLocalStorageConfig(config: IConfig) {
  console.log('config', config);
  const storage = AppStorage.getInstance();
  // game language
  const configLang =
    config.enabledLanguages.length > 0 ? config.enabledLanguages : [LANGUAGE];
  if (!storage.has(EStorageKeys.GAME_LANGUAGE)) {
    storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
  } else {
    // value exists, confirm it is allowed
    if (
      configLang.indexOf(storage.getItem(EStorageKeys.GAME_LANGUAGE)!) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_LANGUAGE, configLang[0]);
    }
  }

  // game word length
  const configLength =
    config.enabledLength.length > 0 ? config.enabledLength : [WORD_LENGTH];
  if (!storage.has(EStorageKeys.GAME_WORD_LENGTH)) {
    storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configLength.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_WORD_LENGTH)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_WORD_LENGTH, `${configLength[0]}`);
    }
  }

  // game attempts
  const configAttempts =
    config.enabledAttempts.length > 0 ? config.enabledAttempts : [ATTEMPTS];
  if (!storage.has(EStorageKeys.GAME_ATTEMPTS)) {
    storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
  } else {
    // value exists, confirm it is allowed
    if (
      configAttempts.indexOf(
        Number(storage.getItem(EStorageKeys.GAME_ATTEMPTS)!),
      ) === -1
    ) {
      storage.setItem(EStorageKeys.GAME_ATTEMPTS, `${configAttempts[0]}`);
    }
  }

  // UI
  if (!storage.has(EStorageKeys.UI_LANGUAGE)) {
    storage.setItem(EStorageKeys.UI_LANGUAGE, LANGUAGE);
  }
}
