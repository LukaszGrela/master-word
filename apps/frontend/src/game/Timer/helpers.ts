import { toHMS } from '@repo/utils';

const prefix0 = (n: number): string => (n < 10 ? `0${n}` : `${n}`);

export const displayTime = (ms: number): string => {
  const { h, m, s } = toHMS(ms);

  if (!h && !m) {
    return `00:${prefix0(s >> 0)}`;
  }

  if (!h && m !== undefined) {
    return `${prefix0(m)}:${prefix0(s)}`;
  }

  if (h !== undefined && m !== undefined) {
    return `${prefix0(h)}:${prefix0(m)}:${prefix0(s)}`;
    /* v8 ignore start */
  }

  return '00:00'; // it should never reach here
};
/* v8 ignore stop */
