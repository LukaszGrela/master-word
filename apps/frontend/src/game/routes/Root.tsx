import { Outlet } from 'react-router-dom';
import { classNames } from '@repo/utils';
import { useLanguage } from '../../i18n';
import Spinner from '../Spinner/Spinner';
import { GrelaDesignIcon } from '../icons/GrelaDesignIcon';
import UILanguage from '../language/UILanguage';

import './Root.scss';

export const Root = () => {
  const { loadedLanguage, loading } = useLanguage();

  const showLoading = loading && !loadedLanguage;

  return (
    <div className={classNames('master-word')}>
      {showLoading && <Spinner />}
      {!showLoading && (
        <>
          <Outlet />
          <p className="footer read-the-docs">
            <GrelaDesignIcon />
            <span>GrelaDesign (c) 2024</span>{' '}
            <span>[v{import.meta.env.VITE_VERSION}]</span>
          </p>
          <UILanguage />
        </>
      )}
    </div>
  );
};
