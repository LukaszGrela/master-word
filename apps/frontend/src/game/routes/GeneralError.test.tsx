import { Mock, afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { defaultLanguageContext } from '../../i18n/LanguageContext/__tests__/contextRenderer';
import GeneralError from './GeneralError';
import { AppStorage, IStorage, hasOwn } from '@repo/utils';
import { IDictionary } from '@repo/common-types';
import { replaceSubstituteMap } from '../../i18n/helpers';
import { createMemoryRouter } from 'react-router-dom';
import { EPaths } from './enums';
import { renderMockRouter } from './__tests__/helpers';
import { Root } from './Root';
import { fireEvent, waitFor } from '@testing-library/react';

vi.mock('@repo/utils', async (importOriginal) => {
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-explicit-any
  const original = await importOriginal<any>();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const AppStorageMock = vi.fn<any, IStorage>();
  const proto = AppStorageMock.prototype as IStorage;

  proto.getItem = vi.fn();
  proto.setItem = vi.fn();
  proto.removeItem = vi.fn();
  proto.has = vi.fn();
  proto.clear = vi.fn();
  proto.clearStorage = vi.fn();

  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const instance = new AppStorageMock();

  // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-member-access
  (AppStorageMock as any).getInstance = vi.fn().mockReturnValue(instance);

  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    ...original,

    AppStorage: AppStorageMock,
  };
});

const originalFetch = global.fetch;

const Boom = () => {
  throw new Error('Boom');
};

describe('game', () => {
  describe('routes', () => {
    describe('GeneralError', () => {
      const translationPl: IDictionary<string> = {
        'translation-info-sr': 'Wybierz język interfejsu',
        'translation-button-en': 'Angielski',
        'translation-button-pl': 'Polski',
        'general-error-title': 'Ożesz kurczaczki!',
        'general-error-message':
          'Przepraszamy, nieoczekiwany błąd jak piorun z nieba spadł i zepsuł zabawę...',
        'general-error-btn-home': 'Zamknij',
      };
      let mockFetch = vi.fn();
      let storage: IStorage = AppStorage.getInstance();
      beforeEach(() => {
        mockFetch = vi.fn();
        storage = AppStorage.getInstance();

        (
          storage.getItem as Mock<[key: string], string | null>
        ).mockImplementation((key) => {
          if (key === 'ui-language') {
            return 'pl';
          }
          if (key === 'word-language') {
            return 'pl';
          }
          return null;
        });

        global.fetch = mockFetch;
      });
      afterEach(() => {
        vi.clearAllMocks();
        global.fetch = originalFetch;
      });
      it('renders general error', async () => {
        const router = createMemoryRouter([
          {
            path: EPaths.ROOT,
            element: <Root />,
            errorElement: <GeneralError />,
            children: [
              {
                index: true,
                element: <p>HOME</p>,
              },
              {
                path: '/boom',
                element: <Boom />,
              },
            ],
          },
        ]);

        await router.navigate('/boom');

        const { container } = renderMockRouter(router, {
          languageContext: {
            ...defaultLanguageContext,
            getUIText: (
              textId: string,
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              substitutes?: IDictionary<any> | undefined,
            ): string => {
              if (hasOwn(translationPl, textId)) {
                const copy = translationPl[textId];
                if (substitutes) {
                  return replaceSubstituteMap(copy, substitutes);
                }
                return copy;
              }

              return textId;
            },
          },
        });

        expect(container).toMatchSnapshot();
      });
      it('close general error', async () => {
        const router = createMemoryRouter([
          {
            path: EPaths.ROOT,
            element: <Root />,
            errorElement: <GeneralError />,
            children: [
              {
                index: true,
                element: <p>HOME</p>,
              },
              {
                path: '/boom',
                element: <Boom />,
              },
            ],
          },
        ]);

        await router.navigate('/boom');

        const { getByText } = renderMockRouter(router, {
          languageContext: {
            ...defaultLanguageContext,
            getUIText: (
              textId: string,
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              substitutes?: IDictionary<any> | undefined,
            ): string => {
              if (hasOwn(translationPl, textId)) {
                const copy = translationPl[textId];
                if (substitutes) {
                  return replaceSubstituteMap(copy, substitutes);
                }
                return copy;
              }

              return textId;
            },
          },
        });

        fireEvent.click(getByText(translationPl['general-error-btn-home']));

        await waitFor(() => {
          expect(router.state.location.pathname).toEqual(EPaths.ROOT);
        });
      });
    });
  });
});
