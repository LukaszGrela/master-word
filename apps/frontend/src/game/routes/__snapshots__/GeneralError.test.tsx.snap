// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`game > routes > GeneralError > renders general error 1`] = `
<div>
  <div
    id="error-page"
  >
    <h1>
      Ożesz kurczaczki!
    </h1>
    <p>
      Przepraszamy, nieoczekiwany błąd jak piorun z nieba spadł i zepsuł zabawę...
    </p>
    <p>
      <i>
        Boom
      </i>
    </p>
    <p>
      <button
        class="primary"
      >
        Zamknij
      </button>
    </p>
    <div
      class="language-selector translation"
    >
      <span
        class="hidden sr"
      >
        Wybierz język interfejsu
      </span>
      <button
        class="translation-btn selected"
        title="Polski"
      >
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/pl.png"
            srcset="https://flagcdn.com/w40/pl.png 2x"
            width="20"
          />
        </span>
      </button>
      <button
        class="translation-btn"
        title="Angielski"
      >
        <span
          class="flag"
        >
          <img
            alt=""
            loading="lazy"
            src="https://flagcdn.com/w20/us.png"
            srcset="https://flagcdn.com/w40/us.png 2x"
            width="20"
          />
        </span>
      </button>
    </div>
  </div>
</div>
`;
